swri_image_util/src/nodelets/replace_colors_nodelet.cpp

```diff
@@ -0,0 +1,457 @@
+      }
+      else
+      {
+        ROS_ERROR("Unknown colormap: %s requested", colormap_name.c_str());
+        success = false;
+      }
+    }
+
+    // Now get the specified number of colors from the requested colormap
+    if (success)
+    {
+      // Make a copy of the grayscale LUT as a working variable
+      cv::Mat original_colors = color_lut_.clone();
+      // color_lut_ will have the transformation from the grayscale values
+      // to the RGB values after this call for every grayscale value
+      cv::applyColorMap(original_colors, color_lut_, colormap_idx);
+
+      // Now modify the orignial colormap to only have the number
+      // of distinct entries specified by the user
+      original_colors = color_lut_.clone();
+
+      int32_t lut_size = color_lut_.cols;
+
+      // Frequently the input image may have some small subset of values,
+      // like 0-5. In this case, just mapping to a colormap will make the 
+      // resulting image look like one color, because the first 6 colors from
+      // the colormap will be used, which for most colormaps are almost the
+      // same value. This will more intelligently remap this values, so that
+      // the 0, 50, 100, 150, 200, 250 color indices are used from the
+      // colormap. This "pushes" the color values apart to make them more
+      // visually apparent.
+      for (int32_t replace_idx = 0; replace_idx < lut_size; replace_idx++)
+      {
+        int32_t lookup_idx = replace_idx % num_entries;
+        lookup_idx = lookup_idx * lut_size / num_entries;
+        color_lut_.at<cv::Vec3b>(0, static_cast<uint8_t>(replace_idx)) = 
+          original_colors.at<cv::Vec3b>(0, static_cast<uint8_t>(lookup_idx));
+      }
+    }
+  }
+
+  // Read the user lookup table from the parameter server
+  void ReplaceColorsNodelet::readUserLut(const XmlRpc::XmlRpcValue& param)
+  {
+    // Assume the parameter parsing works by default
+    bool success = true;
+    if (param.getType() != XmlRpc::XmlRpcValue::TypeArray)
+    {
+      ROS_ERROR("LUT must be an array");
+      success = false;
+    }
+
+    // Make a copy of the current LUT. The copy will be modified, and only
+    // if the complete parameter reading works will the real LUT be modified.
+    cv::Mat temp_lut = color_lut_.clone();
+
+    // Loop over all values that will be replaced. The casting in here is 
+    // complicated because XmlRpc cannot go directly from the parameters to 
+    // uint8_t values, so they must first be cast to int32_t, and then to the
+    // smaller value
+    for (uint32_t lut_idx = 0; success && (lut_idx < param.size()); lut_idx++)
+    {
+      // Each row will be of the form [key, [r, g, b]]
+      XmlRpc::XmlRpcValue lut_row = param[lut_idx];
+      if ((lut_row.getType() != XmlRpc::XmlRpcValue::TypeArray) ||
+        (lut_row.size() != 2))
+      {
+        ROS_ERROR("LUT entries must be two entry arrays");
+        success = false;
+        break;
```

Comment:
Consider not breaking here so that all errors can be found on the first pass, instead of the user potentially getting an error, fixing it, launching again, and getting a different error. Not necessary, but maybe helpful?

Reply:
I thought about that, but given the complexity of the parser it seems like there is a good chance that this could backfire and we end up emitting unhelpful or incorrect error messages because we were trying to parse a badly formatted parameter.